File: Arrays/Count-Subarrays-with-Given-XOR/Python/Count-Subarrays-with-Given-XOR-Optimal.py
from collections import defaultdict

class Solution:
    def subarrayXor(self, arr: list[int], m: int) -> int:
        count = 0
        xr = 0

        freq = defaultdict(int)
        freq[0] = 1

        for num in arr:
            xr ^= num

            if (xr ^ m) in freq:
                count += freq[xr ^ m]

            freq[xr] += 1

        return count

File: Arrays/Count-Subarrays-with-Given-XOR/JavaScript/Count-Subarrays-with-Given-XOR-Optimal.js
var subarrayXor = function(arr, m) {
    let count = 0;
    let xr = 0;

    const freq = new Map();
    freq.set(0, 1);

    for (const num of arr) {
        xr ^= num;

        if (freq.has(xr ^ m)) {
            count += freq.get(xr ^ m);
        }

        freq.set(xr, (freq.get(xr) || 0) + 1);
    }

    return count;
};